File: vane-core/src/main/java/org/oddlama/vane/core/lang/LangMessageArrayField.java
package org.oddlama.vane.core.lang;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Function;
import org.bukkit.configuration.file.YamlConfiguration;
import org.oddlama.vane.annotation.lang.LangMessageArray;
import org.oddlama.vane.core.YamlLoadException;
import org.oddlama.vane.core.module.Module;
import org.oddlama.vane.core.resourcepack.ResourcePackGenerator;

public class LangMessageArrayField extends LangField<TranslatedMessageArray> {

    public LangMessageArray annotation;

    public LangMessageArrayField(
        Module<?> module,
        Object owner,
        Field field,
        Function<String, String> mapName,
        LangMessageArray annotation
    ) {
        super(module, owner, field, mapName);
        this.annotation = annotation;
    }

    @Override
    public void checkLoadable(final YamlConfiguration yaml) throws YamlLoadException {
        checkYamlPath(yaml);

        if (!yaml.isList(yamlPath())) {
            throw new YamlLoadException.Lang("Invalid type for yaml path '" + yamlPath() + "', expected list", this);
        }

        for (final var obj : yaml.getList(yamlPath())) {
            if (!(obj instanceof String)) {
                throw new YamlLoadException.Lang(
                    "Invalid type for yaml path '" + yamlPath() + "', expected string",
                    this
                );
            }
        }
    }

    private List<String> fromYaml(final YamlConfiguration yaml) {
        final var list = new ArrayList<String>();
        for (final var obj : yaml.getList(yamlPath())) {
            list.add((String) obj);
        }
        return list;
    }

    @Override
    public void load(final String namespace, final YamlConfiguration yaml) {
        try {
            field.set(owner, new TranslatedMessageArray(module(), key(), fromYaml(yaml)));
        } catch (IllegalAccessException e) {
            throw new RuntimeException("Invalid field access on '" + field.getName() + "'. This is a bug.");
        }
    }

    @Override
    public void addTranslations(final ResourcePackGenerator pack, final YamlConfiguration yaml, String langCode)
        throws YamlLoadException {
        checkLoadable(yaml);
        final var list = fromYaml(yaml);
        final var loadedSize = get().size();
        if (list.size() != loadedSize) {
            throw new YamlLoadException.Lang(
                "All translation lists for message arrays must have the exact same size. The loaded language file has " +
                loadedSize +
                " entries, while the currently processed file has " +
                list.size(),
                this
            );
        }
        for (int i = 0; i < list.size(); ++i) {
            pack.translations(namespace(), langCode).put(key() + "." + i, list.get(i));
        }
    }
}

File: vane-core/src/main/kotlin/org/oddlama/vane/core/lang/LangMessageArrayField.kt
package org.oddlama.vane.core.lang

import org.bukkit.configuration.file.YamlConfiguration
import org.oddlama.vane.annotation.lang.LangMessageArray
import org.oddlama.vane.core.YamlLoadException
import org.oddlama.vane.core.module.Module
import org.oddlama.vane.core.resourcepack.ResourcePackGenerator
import java.lang.reflect.Field

/**
 * Handles loading and exporting a [TranslatedMessageArray].
 *
 * @param module the owning module.
 * @param owner the target object instance.
 * @param field the reflected target field.
 * @param mapName maps Java field names to YAML paths.
 * @param annotation the source annotation metadata.
 */
class LangMessageArrayField(
    module: Module<*>,
    owner: Any?,
    field: Field,
    mapName: (String?) -> String,
    /** Annotation metadata for this field. */
    var annotation: LangMessageArray?
) : LangField<TranslatedMessageArray?>(module, owner, field, mapName) {

    /**
     * Validates this message array entry in YAML.
     */
    @Throws(YamlLoadException::class)
    override fun checkLoadable(yaml: YamlConfiguration?) {
        if (yaml == null) throw YamlLoadException.Lang("yaml is null", this)
        checkYamlPath(yaml)
        if (!yaml.isList(yamlPath()))
            throw YamlLoadException.Lang("Invalid type for yaml path '${yamlPath()}', expected list", this)
        for (obj in yaml.getList(yamlPath())!!) {
            if (obj !is String)
                throw YamlLoadException.Lang("Invalid type for yaml path '${yamlPath()}', expected string", this)
        }
    }

    /**
     * Reads the message list from YAML.
     */
    private fun fromYaml(yaml: YamlConfiguration): MutableList<String> =
        yaml.getList(yamlPath())!!.map { it as String }.toMutableList()

    /**
     * Loads the translated message array into the reflected field.
     */
    override fun load(namespace: String?, yaml: YamlConfiguration?) {
        if (yaml == null) throw YamlLoadException.Lang("yaml is null", this)
        try {
            field.set(owner, TranslatedMessageArray(module(), key(), fromYaml(yaml)))
        } catch (_: IllegalAccessException) {
            throw RuntimeException("Invalid field access on '${field.name}'. This is a bug.")
        }
    }

    /**
     * Exports this message array to the generated resource pack translations.
     */
    @Throws(YamlLoadException::class)
    override fun addTranslations(pack: ResourcePackGenerator?, yaml: YamlConfiguration?, langCode: String?) {
        if (pack == null || yaml == null || langCode == null) return
        checkLoadable(yaml)
        val list = fromYaml(yaml)
        val loadedSize = get()!!.size()
        if (list.size != loadedSize) {
            throw YamlLoadException.Lang(
                "All translation lists for message arrays must have the exact same size. " +
                        "The loaded language file has $loadedSize entries, while the currently processed file has ${list.size}",
                this
            )
        }
        list.forEachIndexed { i, entry ->
            pack.translations(namespace(), langCode).put("${key()}.$i", entry)
        }
    }
}
